sbi/neural_nets/estimators/categorical_net.py

```diff
@@ -4,13 +4,116 @@
 from typing import Optional
 
 import torch
+from nflows.nn.nde.made import MADE
+from nflows.utils import torchutils
 from torch import Tensor, nn
 from torch.distributions import Categorical
 from torch.nn import Sigmoid, Softmax
+from torch.nn import functional as F
 
 from sbi.neural_nets.estimators.base import ConditionalDensityEstimator
 
 
+class CategoricalMADE(MADE):
+    def __init__(
+        self,
+        categories,  # Tensor[int]
+        hidden_features,
+        context_features=None,
+        num_blocks=2,
+        use_residual_blocks=True,
+        random_mask=False,
+        activation=F.relu,
+        dropout_probability=0.0,
+        use_batch_norm=False,
+        epsilon=1e-2,
+        custom_initialization=True,
+        embedding_net: Optional[nn.Module] = nn.Identity(),
+    ):
+        if use_residual_blocks and random_mask:
+            raise ValueError("Residual blocks can't be used with random masks.")
+
+        self.num_variables = len(categories)
+        self.num_categories = int(max(categories))
+        self.categories = categories
```

Comment:
I am confused by the notion of these three variables. 

what's the difference between variables and categories here?

Reply:
what type does `categories` have?

Reply:
categories lists the number of categories for each categorical variable, i.e. `x_1 = ["heads", "tails"]; x_2 = [1,2,3,4,5,6]` would result in `categories = [2,6]`. I agree the naming might not be ideal. Any suggestions? Also w.r.t. `num_categories`, `num_variables`? Could rename to just `variables` or sth.